File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-go/internal/agente/memoria.go
// Package agente contem as cinco pecas da anatomia de um agente unico
// (Memoria, Planejamento, Ferramentas, Acao, Approval Gate), portadas de
// agent-components-demo.js / agent_components_demo.py.
package agente

// Mensagem e uma mensagem de chat simples {role, content} — equivalente ao
// objeto literal do original.
type Mensagem struct {
	Role    string
	Content string
}

// MemoriaCurtoPrazo vive so durante esta chamada: some quando a funcao
// termina, nunca e reaproveitada pelo proximo protocolo que chegar.
func MemoriaCurtoPrazo(protocolo string) []Mensagem {
	return []Mensagem{
		{Role: "system", Content: "Você redige seções de ICF para estudos clínicos."},
		{Role: "user", Content: "Protocolo: " + protocolo},
	}
}

// Historico e o snapshot imutavel do historico acumulado de um usuario.
type Historico struct {
	Interacoes   int
	Preferencias []string
}

// MemoriaLongoPrazo simula um "banco" simples pra mostrar o principio de
// memoria persistida entre chamadas — em producao seria um banco de dados
// de verdade.
//
// Adaptacao: no original (JS/Python) o "banco" e um Map/dict em escopo de
// modulo, compartilhado por todas as chamadas do processo, e os testes
// chamam .clear() antes de rodar. Aqui isso vira uma struct instanciavel:
// cada MemoriaLongoPrazo tem seu proprio mapa, entao "resetar entre testes"
// e so criar uma instancia nova (NovaMemoriaLongoPrazo()) — mesmo
// comportamento observado, sem depender de estado global mutavel.
type MemoriaLongoPrazo struct {
	bancoDeUsuarios map[string]Historico
}

// NovaMemoriaLongoPrazo cria um banco de memoria de longo prazo vazio.
func NovaMemoriaLongoPrazo() *MemoriaLongoPrazo {
	return &MemoriaLongoPrazo{bancoDeUsuarios: make(map[string]Historico)}
}

// Registrar acumula uma nova interacao (e, opcionalmente, uma preferencia)
// para o usuarioId informado, e devolve o historico atualizado.
func (m *MemoriaLongoPrazo) Registrar(usuarioID string, preferenciaNova string) Historico {
	atual, ok := m.bancoDeUsuarios[usuarioID]
	if !ok {
		atual = Historico{Interacoes: 0, Preferencias: nil}
	}

	novasPreferencias := atual.Preferencias
	if preferenciaNova != "" {
		copia := make([]string, len(atual.Preferencias), len(atual.Preferencias)+1)
		copy(copia, atual.Preferencias)
		novasPreferencias = append(copia, preferenciaNova)
	}

	novo := Historico{
		Interacoes:   atual.Interacoes + 1,
		Preferencias: novasPreferencias,
	}
	m.bancoDeUsuarios[usuarioID] = novo
	return novo
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-java/src/main/java/com/trialforge/agentcomponents/Memoria.java
package com.trialforge.agentcomponents;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 1. MEMORIA — curto prazo (efemera) vs longo prazo (persistida).
 *
 * Porte 1:1 de agent-components-demo.js / agent_components_demo.py (secao 1).
 */
public final class Memoria {

    private Memoria() {
    }

    /** Uma mensagem de chat simples {role, content} — equivalente ao objeto literal do original. */
    public record Mensagem(String role, String content) {
    }

    /**
     * Vive so durante esta chamada: some quando o metodo termina, nunca e reaproveitada
     * pelo proximo protocolo que chegar.
     */
    public static List<Mensagem> memoriaCurtoPrazo(String protocolo) {
        return List.of(
                new Mensagem("system", "Você redige seções de ICF para estudos clínicos."),
                new Mensagem("user", "Protocolo: " + protocolo)
        );
    }

    /** Snapshot imutavel do historico acumulado de um usuario. */
    public record Historico(int interacoes, List<String> preferencias) {
    }

    /**
     * Simula um "banco" simples pra mostrar o principio de memoria persistida entre
     * chamadas — em producao seria um banco de dados de verdade.
     *
     * Adaptacao: no original (JS/Python) o "banco" e um Map/dict em escopo de modulo,
     * compartilhado por todas as chamadas do processo, e os testes chamam .clear() antes
     * de rodar. Aqui isso vira um objeto instanciavel: cada MemoriaLongoPrazo tem seu
     * proprio mapa, entao "resetar entre testes" e so criar uma instancia nova — mesmo
     * comportamento observado, sem depender de estado global mutavel estatico.
     */
    public static class MemoriaLongoPrazo {
        private final Map<String, Historico> bancoDeUsuarios = new ConcurrentHashMap<>();

        public Historico registrar(String usuarioId, String preferenciaNova) {
            Historico atual = bancoDeUsuarios.getOrDefault(usuarioId, new Historico(0, List.of()));
            int novasInteracoes = atual.interacoes() + 1;
            List<String> novasPreferencias = atual.preferencias();
            if (preferenciaNova != null) {
                List<String> copia = new ArrayList<>(atual.preferencias());
                copia.add(preferenciaNova);
                novasPreferencias = List.copyOf(copia);
            }
            Historico novo = new Historico(novasInteracoes, novasPreferencias);
            bancoDeUsuarios.put(usuarioId, novo);
            return novo;
        }
    }
}
